crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -0,0 +1,1367 @@
+            quilt_index.quilt_blocks[i].set_start_index(current_col as u16);
+            current_col += cols_needed;
+            quilt_index.quilt_blocks[i].set_end_index(current_col as u16);
+        }
+
+        let mut final_index_data = Vec::with_capacity(index_total_size);
+        let index_size_u64 = index_total_size as u64;
+        final_index_data.extend_from_slice(&index_size_u64.to_le_bytes());
+        final_index_data
+            .extend_from_slice(&bcs::to_bytes(&quilt_index).expect("Serialization should succeed"));
+
+        // Add the index data to the data.
+        add_blob_to_data(&final_index_data, 0);
+
+        Ok(QuiltV1 {
+            data,
+            row_size,
+            quilt_index,
+            symbol_size,
+        })
+    }
+
+    /// Encodes the blobs into a quilt and returns the slivers.
+    fn encode(&self) -> Result<Vec<SliverPair>, QuiltError> {
+        let _guard = self.span.enter();
+        tracing::event!(Level::DEBUG, "starting to encode quilt");
+
+        let quilt = self.construct_quilt()?;
+        let encoder = BlobEncoder::new(self.config.clone(), quilt.data()).map_err(|_| {
+            QuiltError::quilt_oversize(format!("quilt is too large: {}", quilt.data().len()))
+        })?;
+        assert_eq!(encoder.symbol_usize(), quilt.symbol_size());
+        Ok(encoder.encode())
+    }
+
+    /// Encodes the blobs into a quilt and returns the slivers and metadata.
+    fn encode_with_metadata(&self) -> Result<(Vec<SliverPair>, QuiltMetadata), QuiltError> {
+        let _guard = self.span.enter();
+        tracing::debug!("starting to encode quilt with metadata");
+
+        let quilt = self.construct_quilt()?;
+        let encoder = BlobEncoder::new(self.config.clone(), quilt.data()).map_err(|_| {
+            QuiltError::quilt_oversize(format!("quilt is too large: {}", quilt.data.len()))
+        })?;
+
+        assert_eq!(encoder.symbol_usize(), quilt.symbol_size);
+
+        let (sliver_pairs, metadata) = encoder.encode_with_metadata();
+        let quilt_metadata = QuiltMetadata {
+            quilt_id: *metadata.blob_id(),
+            metadata: metadata.metadata().clone(),
+            index: quilt.quilt_index().clone(),
+        };
+
+        Ok((sliver_pairs, quilt_metadata))
+    }
+}
+
+impl<'a> QuiltEncoderV1<'a> {
+    /// Creates a new [`QuiltEncoderV1`] from a encoding config and a set of blobs.
+    pub fn new(config: EncodingConfigEnum<'a>, blobs: &'a [BlobWithIdentifier<'a>]) -> Self {
+        Self {
+            blobs,
+            config,
+            span: tracing::span!(Level::ERROR, "QuiltEncoderV1"),
+        }
+    }
+}
+
+/// Finds the minimum symbol size needed to store blobs in a fixed number of columns.
+/// Each blob must be stored in consecutive columns exclusively.
+///
+/// The first column is used to store the quilt index, so the symbol size has to be
+/// large enough to store the quilt index in a single column.
```

Comment:
Important: Above you state that the index can occupy one or multiple columns. Can you clarify which is correct?